clients/drcachesim/tools/opcode_mix.cpp

```diff
@@ -222,72 +203,24 @@ opcode_mix_t::parallel_shard_memref(void *shard_data, const memref_t &memref)
-        decode_pc = const_cast<app_pc>(memref.instr.encoding);
-        if (memref.instr.encoding_is_new) {
-            // The code may have changed: invalidate the cache.
-            shard->worker->opcode_data_cache.erase(trace_pc);
-        }
-    } else {
-        // Legacy trace support where we need the binaries.
-        if (!module_mapper_) {
-            shard->error =
-                "Module file path is missing and trace has no embedded encodings";
+    } else if (shard->decode_cache == nullptr) {
+        std::string err = init_decode_cache(shard, dcontext_.dcontext, shard->filetype);
+        if (err != "") {
+            shard->error = err;
             return false;
         }
-        if (trace_pc >= shard->last_trace_module_start &&
-            static_cast<size_t>(trace_pc - shard->last_trace_module_start) <
-                shard->last_trace_module_size) {
-            decode_pc = shard->last_mapped_module_start +
-                (trace_pc - shard->last_trace_module_start);
-        } else {
-            std::lock_guard<std::mutex> guard(mapper_mutex_);
-            decode_pc = module_mapper_->find_mapped_trace_bounds(
-                trace_pc, &shard->last_mapped_module_start,
-                &shard->last_trace_module_size);
-            if (!module_mapper_->get_last_error().empty()) {
-                shard->last_trace_module_start = nullptr;
-                shard->last_trace_module_size = 0;
-                shard->error = "Failed to find mapped address for " +
-                    to_hex_string(memref.instr.addr) + ": " +
-                    module_mapper_->get_last_error();
-                return false;
-            }
-            shard->last_trace_module_start =
-                trace_pc - (decode_pc - shard->last_mapped_module_start);
-        }
     }
-    int opcode;
-    uint category;
-    auto cached_opcode_category = shard->worker->opcode_data_cache.find(trace_pc);
-    if (cached_opcode_category != shard->worker->opcode_data_cache.end()) {
-        opcode = cached_opcode_category->second.opcode;
-        category = cached_opcode_category->second.category;
-    } else {
-        instr_t instr;
-        instr_init(dcontext_.dcontext, &instr);
-        app_pc next_pc =
-            decode_from_copy(dcontext_.dcontext, decode_pc, trace_pc, &instr);
-        if (next_pc == NULL || !instr_valid(&instr)) {
-            instr_free(dcontext_.dcontext, &instr);
-            shard->error =
-                "Failed to decode instruction " + to_hex_string(memref.instr.addr);
-            return false;
-        }
-        opcode = instr_get_opcode(&instr);
-        category = instr_get_category(&instr);
-        shard->worker->opcode_data_cache[trace_pc] = opcode_data_t(opcode, category);
-        instr_free(dcontext_.dcontext, &instr);
+
+    ++shard->instr_count;
+    opcode_data_t *opcode_data;
+    shard->error = shard->decode_cache->add_decode_info(memref.instr, opcode_data);
```

Comment:
May be worth profiling an opcode_mix analyzer run on a medium-sized trace on local disk just to confirm there's no caching/lookup bottleneck we don't expect.

Reply:
Done, using an opcode_mix run on a trace with embedded encodings and one without.

It spends about the same cpu time (somewhat lower in the trace with embedded encodings) for caching, with and without this change.

I'm seeing a 3x higher memory usage though in the cache unordered_map. It could be because now the cache has been moved to the shard, instead of the worker, but we do a clear at the end of each shard. I'm still checking if this is an artifact of how I measured the memory usage.

Reply:
This was actually a performance bug. std::unordered_map does not release all of its internal memory on a clear(). So despite doing a clear_cache() for the shard's cache in parallel_shard_exit(), some memory was still stuck allocated. Fixed in [b9c0fef](https://github.com/DynamoRIO/dynamorio/pull/7114/commits/b9c0fefc09a1bb6b6a02fca592853da735bd1343).

Reply:
Made a simpler fix in a10721bb306bd92d4fc53af06914a50921f87fab instead.